README
NAME
    Alien::IUP - Building, finding and using iup + related libraries -
    <http://www.tecgraf.puc-rio.br/iup/>

VERSION
    Version 0.400_6 of Alien::IUP is based on the following:

    *   *iup* library 3.6 - see <http://www.tecgraf.puc-rio.br/iup/>

    *   *im* library 3.8 - see <http://www.tecgraf.puc-rio.br/im/>

    *   *cd* library 5.5.1 - see <http://www.tecgraf.puc-rio.br/cd/>

SYNOPSIS
    IMPORTANT: This module is not a perl binding for *iup + related*
    libraries; it is just a helper module. The real perl binding is
    implemented by IUP module, which is using Alien::IUP to locate *iup +
    related* libraries on your system (or build it from source codes).

    Alien::IUP tries (in given order) during its installation:

    *   Locate an already installed *iup* and related libraries + ask user
        whether to use the already installed *iup* or whether to build *iup*
        from sources

    *   Via env variable IUP_DIR you can specify where the build script
        should look for the already installed *iup* and related libs
        (directories $ENV{IUP_DIR}/lib and $ENV{IUP_DIR}/include are
        expected to exist)

    *   When not using the already installed libraries build process
        continues with the following steps

    *   Download *iup* & co. source code tarballs

    *   Build *iup* & co. binaries from source codes (note: static libraries
        are build in this case)

    *   Install libs and dev files (*.h, *.a) into *share* directory of
        Alien::IUP distribution - *share* directory is usually something
        like this: /usr/lib/perl5/site_perl/5.10/auto/share/dist/Alien-IUP

    Later on you can use Alien::IUP in your module that needs to link with
    *iup* and/or related libs like this:

     # Sample Makefile.pl
     use ExtUtils::MakeMaker;
     use Alien::IUP;
 
     WriteMakefile(
       NAME         => 'Any::IUP::Module',
       VERSION_FROM => 'lib/Any/IUP/Module.pm',
       LIBS         => Alien::IUP->config('LIBS'),
       INC          => Alien::IUP->config('INC'),
       # + additional params
     );

    IMPORTANT: As Alien::IUP builds static libraries the modules using
    Alien::IUP (e.g. IUP) need to have Alien::IUP just for building, not for
    later use. In other words Alien:IUP is just "build dependency" not
    "run-time dependency".

METHODS
  config()
    This function is the main public interface to this module.

     Alien::IUP->config('LIBS');

    Returns a string like: '-L/path/to/iupdir/lib -liup -lim -lcd'

     Alien::IUP->config('INC');

    Returns a string like: '-I/path/to/iupdir/include'

     Alien::IUP->config('PREFIX');

    Returns a string like: '/path/to/iupdir' (note: if using the already
    installed *iup* config('PREFIX') returns undef)

  havelib()
    [since 0.115] Checks the presence of given iup related libraries.

     Alien::IUP->havelib('iupim');
     #or
     Alien::IUP->havelib('iupim', 'iupcd', 'iupcontrols');

    Parameter(s): One or more iup related lib names - e.g. iup, cd, im,
    iupcd, iupim, iupcontrols, iup_pplot, iupimglib, iupgl, iupole.

    Returns: 1 if all libs specified as a param are available; 0 otherwise.

AUTHOR
    KMX, <kmx at cpan.org>

BUGS
    Please report any bugs or feature requests to "bug-alien-iup at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-IUP>.

LICENSE AND COPYRIGHT
    Libraries *iup*, *im* and *cd*: Copyright (C) 1994-2010 Tecgraf,
    PUC-Rio. <http://www.tecgraf.puc-rio.br>

    Alien::IUP module: Copyright (C) 2010 KMX.

    This program is distributed under the MIT License:
    <http://www.opensource.org/licenses/mit-license.php>

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
=head1 Revision history for Alien-IUP

=head2 14/08/2012 - 0.400_6

- IMPORTANT: switch to iup-3.6 (new iup_mglplot introduced)

=head2 20/08/2011 - 0.305

 - iup-imglib patch
 - cd patch (crutial - without this patch IUP module segfaults at some points)

=head2 19/08/2011 - 0.304

 - cygwin patch (__argc + __argv related)

=head2 08/08/2011 - 0.303

 - added support for -DIUP_IMGLIB_LARGE (now on Win32 only)
 - fixed http://www.cpantesters.org/cpan/report/889253e2-b5f8-11e0-8204-0f92f06c0dc8 (kFreeBSD failure)
 - better diagnostic messages

=head2 22/07/2011 - 0.302

 - solaris workaround (ar, ranlib)

=head2 12/05/2011 - 0.301

 - IMPORTANT: switch to iup-3.5

=head2 05/04/2011 - 0.204

 - IMPORTANT: the last release with iup-3.4
 - back from Digest::file to Digest::SHA

=head2 04/04/2011 - 0.203

 - fixed "missing download dir" issue

=head2 03/04/2011 - 0.202

 - fixed http://www.cpantesters.org/cpan/report/61d37376-471c-11e0-b0ff-ff1c7a0d71a6 (File::Spec & devnull)
 - using LWP::Simple instead of File::Fetch for download
 - using Digest::file instead of Digest::SHA

=head2 18/02/2011 - 0.201

 - IMPORTANT: the first stable release with iup-3.4
 - dev build: switch to iup-3.4 + updated im, cd

=head2 10/02/2011 - 0.116

 - IMPORTANT: the last release with iup-3.3
 - Unix: if not enough libs found - show more debug info

=head2 03/02/2011 - 0.115

 - Alien::IUP->havelib() introduced
 - preinstalled lib detection disabled (untill somebody will ask for it)

=head2 27/01/2011 - 0.114

 - not building libs that are currently not used by IUP
 - less build messages by default
 - dev build: with the latest CVS sources

=head2 05/01/2011 - 0.113

 - same as 0.112_01 + updated TODO
 - fixing -lstdc++ issue on non-gcc compilers

=head2 30/12/2010 - 0.112

 - skipping iuptuio on solaris - http://www.cpantesters.org/cpan/report/c39d9264-0ae3-11e0-afb0-adca6bb533f3
 - skipping iuptuio on cygwin (not only gcc3)

=head2 16/12/2010 - 0.111

 - GTK/GTK_BASE workaround fixing http://www.cpantesters.org/cpan/report/d67a6578-08aa-11e0-afb0-adca6bb533f3

=head2 13/12/2010 - 0.110

 - iup-gtk-tabs.diff fixing http://www.cpantesters.org/cpan/report/b524182a-04c1-11e0-afb0-adca6bb533f3
 - fixing gtkprintunixdialog.h failure http://www.cpantesters.org/cpan/report/9629bbba-0655-11e0-9ee0-f20c722eace1

=head2 12/12/2010 - 0.109

 - fixing im-INC-order.diff + cd-GTK_BASE.diff

=head2 09/12/2010 - 0.108_01+0.108_02

 - dev build: with the latest CVS sources

=head2 08/12/2010 - 0.108

 - info.t fix

=head2 07/12/2010 - 0.107

 - iuptuio target added (Unix.pm)
 - more debug info saved to config_data
 - updated im-INC-order.diff

=head2 02/12/2010 - 0.106

 - updated cd-GTK_BASE.diff

=head2 02/12/2010 - 0.105

 - important fix: GTK2 + pkg-config
 - improved build target autoselection

=head2 01/12/2010 - 0.104

 - updated iup-GTK_WIDGET.diff

=head2 01/12/2010 - 0.103

 - Motif/X11 build fix
 - fix for http://www.cpantesters.org/cpan/report/ec6300d2-fb91-11df-a694-e2b03d723d5a

=head2 01/12/2010 - 0.102

 - patches: cd-GTK_BASE.diff, iup-GTK_WIDGET.diff
 - linking lib order fixed
 - showing mod version in debug info

=head2 23/11/2010 - 0.101

 - IMPORTANT: switch to IUP 3.3 (+ upgrade of other libs as well)
 - NetBSD ntohl() related patch (#include <sys/types.h>)
 - significant Windows build changes
 - warn (not skip) by im_process on OpenBSD
 - do not build freetype if we halready have freetype on our system
 - ms64bit build patch
 - more diagnostics (freetype)
 - require installed cairo for GTK2 build
 - fixing netbsd failure (using USE_PKGCONFIG)
 - webkit/webkit.h test (disabling iupweb if no webkit)
 - patches for DragonFly BSD
 - improved cygwin build
 - new checksum for updated iup-3.3_Sources.tar.gz
 - new builder for cygwin (separated from generic UNIX)

=head2 14/11/2010 - 0.024

 - IMPORTANT: the last release with iup-3.2
 - die if no GUI lib detected
 - switching Changes fromat from plain text to POD
 - changing version numbering

=head2 20/10/2010 - 0.0.23

 - create_build_script moved after GNU make check

=head2 12/10/2010 - 0.0.22

 - fixed GNU make check exit code

=head2 11/10/2010 - 0.0.21

 - GNU make check moved to Build.pl (now cpan testers should report "N/A)
 - more messages on sha1 check failure
 - small POD changes

=head2 05/10/2010 - 0.0.20

 - more questions/prompts during build
 - cleaning up + cosmetics
 - OpenBSD - not building im_process (g++ internal error)
 - OpenBSD - lesstif related fixes
 - DragonFly patch for 'sizeof(FILE)'
 - exit if GNU make not found
 - DragonFly patch for libjasper/jasper/jas_types.h
 - create_makefile_pl => 'small' (creating Makefile.PL to satisfy old CPAN clients)

=head2 28/09/2010 - 0.0.19

 - iup patch - fixing newlines in the main Makefile (troubles on solaris)

=head2 24/09/2010 - 0.0.18

 - using File::Path legacy functions
 - fixed gtk2 - freetype2 collision (tecmake.mak)

=head2 22/09/2010 - 0.0.17

 - fix for cygwin CPAN Tester failure
 - cleaning sharedir during install

=head2 21/09/2010 - 0.0.16

 - using pkg-config in tecmake.mak

=head2 17/09/2010 - 0.0.15

 - stdc++ fix (needed by Linux)

=head2 16/09/2010 - 0.0.14

 - cygwin+Win64 fixes
 - removing testing pod

=head2 15/09/2010 - 0.0.13

 - fixing liborder
 - if "No GUI lib" then defaulting to X11
 - another POD test

=head2 25/08/2010 - 0.0.12

 - GTK vs. GTK_BASE fix
 - printing output of make commands only on failure

=head2 23/08/2010 - 0.0.11

 - GTK_BASE detection fix

=head2 18/08/2010 - 0.0.10

 - added MS Win build (gcc, msvc)
 - improved lib/inc dir detection (unix)
 - building static libs
 - using IPC::Run3 for launching external commands

=head2 05/08/2010 - 0.0.9

 - FAIL ALWAYS debug version
 - fixed bug in setting GTK make param

Build.PL
use 5.008;
use strict;
use warnings;
use Module::Build;

use lib 'inc';

my %platforms =(
  # Unix = default, thus not listing all UNIX like systems
  MSWin32 => 'Windows',
  cygwin  => 'Cygwin',
  linux   => 'Unix',
);
my $package = 'My::Builder::' . ($platforms{$^O} || 'Unix');
print STDERR "Gonna use '$package' class ...\n";
eval "require $package" or die "###ERROR### Require '$package' failed: $@\n";

my $builder = $package->new(
  module_name        => 'Alien::IUP',
  dist_version_from  => 'lib/Alien/IUP.pm',
  dist_author        => 'KMX, kmx@cpan.org',
  license            => 'mit',
  requires => {
    'perl'              => 5.008,
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
  },
  configure_requires => {
    'perl'              => 5.008,
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
    'File::Path'        => 0,
    'File::Temp'        => 0,
    'File::Find'        => 0,
    'ExtUtils::CBuilder'=> 0,
    'ExtUtils::Command' => 0,
    'ExtUtils::Liblist' => 0,
    'File::Fetch'       => 0,
    'Digest::SHA'       => 0,
    'Archive::Extract'  => 0,
    'Text::Patch'       => 0,
    'IPC::Run3'         => 0.041, # we need return_if_system_error
  },
  build_requires => {
    'perl'              => 5.008,
    'Test::More'        => 0,
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
    'File::Path'        => 0,
    'File::Temp'        => 0,
    'File::Find'        => 0,
    'ExtUtils::CBuilder'=> 0,
    'ExtUtils::Command' => 0,
    'ExtUtils::Liblist' => 0,
    'File::Fetch'       => 0,
    'Digest::SHA'       => 0,
    'Archive::Extract'  => 0,
    'Text::Patch'       => 0,
    'IPC::Run3'         => 0.041, # we need return_if_system_error
  },
  share_dir          => 'sharedir',
  add_to_cleanup     => [ 'Alien-IUP-*', 'build_done', 'build_src' ],
  #create_makefile_pl => 'small', # creating Makefile.PL to satisfy old CPAN clients (does not work well with our style of M::B use)
  create_readme      => 1,
  meta_merge => {
    resources  => {
      bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-IUP',
      repository => 'http://github.com/kmx/alien-iup',
    }
  },
);

unlink 'build_done'; # delete build marker (= force rebuild after running Build.PL)

if ($builder->check_installed_lib()) {
  my $ans = $builder->prompt("\nIUP detected, wanna use IUP already installed on your system (y/n)?", 'y');
  $builder->notes('already_installed_lib', undef) unless (lc($ans) eq 'y');  
}

$builder->notes('build_debug_info', lc($builder->prompt("\nWanna see more debug info during build (y/n)?", 'y')) eq 'y' ? 1 : 0);

# set what tarballs we are gonna use
$builder->notes('iup_url', 'http://download.sf.net/iup/iup-3.3_Sources.tar.gz');
$builder->notes('iup_sha1', '6ce0badbad9e27131010bf6fd296b1d5c8eda0a8');
$builder->notes('iup_patches', [ 'patches/iup-3.3.diff' ] );

$builder->notes('im_url', 'http://download.sf.net/imtoolkit/im-3.6.3_Sources.tar.gz');
$builder->notes('im_sha1', '93f5e1fef25d9ac42a8c3a628411156b695cd7d1');
$builder->notes('im_patches', [ 'patches/im-3.6.3.diff' ] );

$builder->notes('cd_url', 'http://download.sf.net/canvasdraw/cd-5.4.1_Sources.tar.gz');
$builder->notes('cd_sha1', 'b771752f253ab3bf6ee8c4bea228d5d16f4352ec');
#$builder->notes('cd_patches', [ 'patches/cd.diff' ] );

# check if we have GNU make otherwise 'exit 0' - cpan testers will report "N/A"
my $m = $builder->get_make;
exit 0 unless $m;
$builder->notes('gnu_make', $m);

$builder->create_build_script();


inc/My/Builder/Cygwin.pm
package My::Builder::Cygwin;

use strict;
use warnings;
use base 'My::Builder';

use File::Spec::Functions qw(catdir catfile rel2abs);
use File::Spec qw(devnull);
use File::Glob qw(glob);
use File::Copy;
use Config;

sub build_binaries {
  my ($self, $build_out, $srcdir) = @_;
  my $success = 1;
  
  #do the job
  $success = $self->build_via_tecmake($build_out, $srcdir);
  warn "###MAKE FAILED###" unless $success;

  #make a list of libs necessary to link with IUP and related libraries
  my %seen;
  my @gl_l = glob("$build_out/lib/*");
  my @gl_i = glob("$build_out/include/*");
  print STDERR "Output counts: lib=" . scalar(@gl_l) . " include=" . scalar(@gl_i) . "\n";
  if ((scalar(@gl_l) < 3) || (scalar(@gl_i) < 3)) {
    warn "###WARN### $build_out/lib/ or $build_out/include/ not complete";
    $success = 0;
  }
  foreach (@gl_l) {
    if ($_ =~ /lib([a-zA-Z0-9\_\-\.]*?)\.(a|dll\.a)$/) {
      $seen{$1} = 1;
    }
    elsif ($_ !~ /\.dll$/) { # *.dll on cygwin is OK
      warn "###WARN### Unexpected filename '$_'";
      $success = 0;
    }
  }
  print STDERR "Output libs: $_\n" foreach (sort keys %seen);
  my @libs = $self->sort_libs(keys %seen);

  $self->config_data('extra_cflags', '');
  $self->config_data('extra_lflags', '');
  $self->config_data('linker_libs', [ @libs, qw/gdi32 comdlg32 comctl32 winspool uuid ole32 oleaut32 opengl32 glu32/ ] );

  die "###BUILD ABORTED###" unless $success;
  print STDERR "Build finished sucessfully!\n";
};

sub build_via_tecmake {
  my ($self, $build_out, $srcdir) = @_;
  my $prefixdir = rel2abs($build_out);

  my @cdtgs = qw[cd_freetype cd_ftgl config cd_pdflib cdpdf cdgl]; #config = default target (libcd)
  my @imtgs = qw[config im_process im_jp2 im_fftw]; #config = default target (libim)
  my @iuptgs = qw[src srccd srccontrols srcpplot srcgl srcim srcimglib srcole]; #!!! srcweb fails

  my $im_si;
  my $success = 1;

  # save it for future use in ConfigData
  $self->config_data('build_prefix', $prefixdir);

  #create output directory structure
  mkdir "$prefixdir" unless -d "$prefixdir";
  mkdir "$prefixdir/lib" unless -d "$prefixdir/lib";
  mkdir "$prefixdir/include" unless -d "$prefixdir/include";

  my %done;
  my $tecuname = 'gcc4';
  #my $tecuname = 'dllg4';
  my @basecmd = (qw[make -f ../tecmakewin.mak USE_NODEPEND=Yes], "TEC_UNAME=$tecuname");

  if(-d "$srcdir/im/src") {
    print STDERR "Gonna build 'im'\n";
    chdir "$srcdir/im/src";
    # some debug info
    $im_si = $self->run_output_std(@basecmd, 'sysinfo') if $self->notes('build_debug_info');    
    foreach my $t (@imtgs) {
      $done{"im:$t"} = $self->run_custom(@basecmd, "MF=$t");
      $success = 0 unless $done{"im:$t"};
    }
    copy($_, "$prefixdir/include/") foreach (glob("../include/*.h"));
    copy($_, "$prefixdir/lib/") foreach (glob("../lib/$tecuname/*"));
    chdir $self->base_dir();
  }

  if (-d "$srcdir/cd/src") {
    print STDERR "Gonna build 'cd'\n";
    chdir "$srcdir/cd/src";
    foreach my $t (@cdtgs) {
      $done{"cd:$t"} = $self->run_custom(@basecmd, "MF=$t");
      $success = 0 unless $done{"cd:$t"};
    }
    copy($_, "$prefixdir/include/") foreach (glob("../include/*.h"));
    copy($_, "$prefixdir/lib/") foreach (glob("../lib/$tecuname/*"));
    chdir $self->base_dir();
  }

  if (-d "$srcdir/iup") {
    print STDERR "Gonna build 'iup'\n";    
    foreach my $t (@iuptgs) {
      print STDERR "changing dir '$srcdir/iup/$t'\n";    
      chdir "$srcdir/iup/$t";
      $done{"iup:$t"} = $self->run_custom(@basecmd);
      $success = 0 unless $done{"iup:$t"};
      chdir $self->base_dir();
    }
    chdir "$srcdir/iup";
    copy($_, "$prefixdir/include/") foreach (glob("./include/*.h"));
    copy($_, "$prefixdir/lib/") foreach (glob("./lib/$tecuname/*"));
    chdir $self->base_dir();
  }

  print STDERR "Done: $done{$_} - $_\n" foreach (sort keys %done);
  $self->config_data('debug_done', \%done);
  $self->config_data('debug_si', $im_si);
  
  return $success;
}

sub get_make {
  # on cygwin always 'make'
  return 'make';
}

sub quote_literal {
    my ($self, $txt) = @_;
    $txt =~ s|'|'\\''|g;
    return "'$txt'";
}

1;


lib/Alien/IUP.pm
package Alien::IUP;

use warnings;
use strict;

use Alien::IUP::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec::Functions qw(catdir catfile rel2abs);

=head1 NAME

Alien::IUP - Building, finding and using iup + related libraries - L<http://www.tecgraf.puc-rio.br/iup/>

=cut

# following recommendation from http://www.dagolden.com/index.php/369/version-numbers-should-be-boring/
our $VERSION = "0.400_6";
$VERSION = eval $VERSION;

=head1 VERSION

Version 0.400_6 of Alien::IUP is based on the following:

=over

=item * I<iup> library 3.6 - see L<http://www.tecgraf.puc-rio.br/iup/>

=item * I<im> library 3.8 - see L<http://www.tecgraf.puc-rio.br/im/>

=item * I<cd> library 5.5.1 - see L<http://www.tecgraf.puc-rio.br/cd/>

=back

=head1 SYNOPSIS

B<IMPORTANT:> This module is not a perl binding for I<iup + related> libraries; it is just
a helper module. The real perl binding is implemented by L<IUP|IUP> module,
which is using Alien::IUP to locate I<iup + related> libraries on your system (or build it from source codes).

Alien::IUP tries (in given order) during its installation:

=over

=item * Locate an already installed I<iup> and related libraries  + ask user whether
to use the already installed I<iup> or whether to build I<iup> from sources

=item * Via env variable IUP_DIR you can specify where the build script should look
for the already installed I<iup> and related libs (directories $ENV{IUP_DIR}/lib and 
$ENV{IUP_DIR}/include are expected to exist)

=item * When not using the already installed libraries build process continues with
the following steps

=item * Download I<iup> & co. source code tarballs

=item * Build I<iup> & co. binaries from source codes (note: static libraries are build in this case)

=item * Install libs and dev files (*.h, *.a) into I<share> directory of Alien::IUP
distribution - I<share> directory is usually something like this: /usr/lib/perl5/site_perl/5.10/auto/share/dist/Alien-IUP

=back

Later on you can use Alien::IUP in your module that needs to link with
I<iup> and/or related libs like this:

 # Sample Makefile.pl
 use ExtUtils::MakeMaker;
 use Alien::IUP;
 
 WriteMakefile(
   NAME         => 'Any::IUP::Module',
   VERSION_FROM => 'lib/Any/IUP/Module.pm',
   LIBS         => Alien::IUP->config('LIBS'),
   INC          => Alien::IUP->config('INC'),
   # + additional params
 );

B<IMPORTANT:> As Alien::IUP builds static libraries the modules using Alien::IUP (e.g. L<IUP|IUP>)
need to have Alien::IUP just for building, not for later use. In other words Alien:IUP is just
"build dependency" not "run-time dependency".

=head1 METHODS

=head2 config()

This function is the main public interface to this module.

 Alien::IUP->config('LIBS');

Returns a string like: '-L/path/to/iupdir/lib -liup -lim -lcd'

 Alien::IUP->config('INC');

Returns a string like: '-I/path/to/iupdir/include'

 Alien::IUP->config('PREFIX');

Returns a string like: '/path/to/iupdir' (note: if using the already installed
I<iup> config('PREFIX') returns undef)

=head2 havelib()

[since 0.115] Checks the presence of given iup related libraries.

 Alien::IUP->havelib('iupim');
 #or
 Alien::IUP->havelib('iupim', 'iupcd', 'iupcontrols');

Parameter(s): One or more iup related lib names - e.g. iup, cd, im, iupcd, iupim, iupcontrols, iup_pplot, iupimglib, iupgl, iupole.

Returns: 1 if all libs specified as a param are available; 0 otherwise.

=head1 AUTHOR

KMX, E<lt>kmx at cpan.orgE<gt>

=head1 BUGS

Please report any bugs or feature requests to C<bug-alien-iup at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-IUP>.

=head1 LICENSE AND COPYRIGHT

Libraries I<iup>, I<im> and I<cd>: Copyright (C) 1994-2010 Tecgraf, PUC-Rio.
L<http://www.tecgraf.puc-rio.br>

Alien::IUP module: Copyright (C) 2010 KMX.

This program is distributed under the MIT License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut

sub config {
  my ($package, $param) = @_;
  return unless ($param =~ /[a-z0-9_]*/i);
  my $subdir = Alien::IUP::ConfigData->config('share_subdir');
  unless ($subdir) {
    #we are using lib already installed on your system not compiled by Alien
    #therefore no additinal magic needed
    return Alien::IUP::ConfigData->config('config')->{$param};
  }
  my $share_dir = dist_dir('Alien-IUP');
  my $real_prefix = catdir($share_dir, $subdir);
  my $val = Alien::IUP::ConfigData->config('config')->{$param};
  return unless $val;
  $val =~ s/\@PrEfIx\@/$real_prefix/g; # handle @PrEfIx@ replacement
  return $val;
}

sub havelib {
  my ($package, @libs) = @_;
  my $iuplibs = my $subdir = Alien::IUP::ConfigData->config('iup_libs');
  return 0 unless defined $iuplibs;
  for (@libs) {
    return 0 unless defined $iuplibs->{$_} && $iuplibs->{$_} == 1;
  }
  return 1;
}

1; # End of Alien::IUP
